punchcut/src/outline.rs

```diff
@@ -0,0 +1,293 @@
+/*!
+Representation of a simple outline glyph.
+*/
+
+use peniko::kurbo::{
+    segments, BezPath, ParamCurveArclen, ParamCurveArea, ParamCurveExtrema, PathEl, Point, Rect,
+    Shape,
+};
+
+use crate::source::*;
+
+/// Simple outline represented as a sequence of path commands.
+#[derive(Clone, Default, Debug)]
+pub struct Outline {
+    verbs: Vec<Verb>,
+    points: Vec<(f32, f32)>,
+}
+
+impl Outline {
+    /// Creates a new empty outline.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Creates a new outline from a TrueType outline. Returns `None` if
+    /// the source outline is malformed.
+    pub fn from_glyf(outline: &glyf::Outline) -> Option<Self> {
+        let mut result = Self::new();
+        if result.append_glyf(outline) {
+            Some(result)
+        } else {
+            None
+        }
+    }
+
+    /// Returns true if the outline is empty.
+    pub fn is_empty(&self) -> bool {
+        self.verbs.is_empty()
+    }
+
+    /// Resets the outline to the empty state.
+    pub fn clear(&mut self) {
+        self.verbs.clear();
+        self.points.clear();
+    }
+
+    /// Returns an iterator over the path elements of the outline.
+    pub fn elements(&self) -> Elements {
+        Elements {
+            verbs: &self.verbs,
+            points: &self.points,
+            verb_pos: 0,
+            point_pos: 0,
+        }
+    }
+
+    /// Appends a glyf outline, converting it into a sequence of path elements.
+    /// Returns false if the source outline is malformed.
+    pub fn append_glyf(&mut self, outline: &glyf::Outline) -> bool {
+        #[inline(always)]
+        fn conv(p: glyf::Point, s: f32) -> (f32, f32) {
```

Comment:
maybe worth just adding an impl in `read-fonts`.

Reply:
It might make sense to change the `Point` struct in `read-fonts` to take a type parameter and then potentially add a "convert to f32 with scale method"